pkg/mutation/system.go

```diff
@@ -0,0 +1,165 @@
+package mutation
+
+import (
+	"fmt"
+	"sort"
+	"sync"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation/mutator"
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation/schema"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+// System keeps the list of mutations and
+// provides an interface to apply mutations.
+type System struct {
+	schemaDB schema.DB
+	mutators []mutator.Mutator
+	sync.RWMutex
+}
+
+// NewSystem initializes an empty mutation system
+func NewSystem() *System {
+	return &System{
+		mutators: make([]mutator.Mutator, 0),
+	}
+}
+
+// Upsert updates or insert the given object, and returns
+// an error in case of conflicts
+func (s *System) Upsert(m mutator.Mutator) error {
+	s.Lock()
+	defer s.Unlock()
+
+	// Checking schema consistence only if the mutator has schema
+	if withSchema, ok := m.(mutator.WithSchema); ok {
```

Comment:
We'll want to check for diffs and exit early before doing anything else if there are no diffs.